Fix for torch.inf.to(torch.int32) produces different values on CPU vs…

vishalgoyal316 opened the pull request:
Code Changes for:

Mapping from +inf to max, -inf to min and Nan to 0, when we type cast to integer type.
Updating for respective test case.

Fixes #154311

vishalgoyal316 committed:
Fix for torch.inf.to(torch.int32) produces different values on CPU vs CUDA Issue#154311

vishalgoyal316 commented:
@pytorchbot label "topic: not user facing"

malfet reviewed (commented):
What are the performance implications? And why this is needed? 
I.e. currently it's an undefined behavior per all language specs, why it  needs to be standartized?

malfet reviewed (changes requested):
Please specify how to do you measure/estimate performance impact of this change
Also, this will requires a documentation update, as it used to be an undefined behavior, but now it is getting specified.

malfet commented on test/test_tensor_creation_ops.py:
```diff
@@ -1086,7 +1086,7 @@ def test_float_to_int_conversion_finite(self, device, dtype):
     # nondeterministically fails, warning "invalid value encountered in cast"
     @onlyCPU
```
If you want this behavior to be uniform, why limit it to `onlyCPU`?

vishalgoyal316 committed:
removed @onlyCPU from testcase of test_float_to_int_conversion_nonfinite for CUDA

vishalgoyal316 committed:
Document well-defined behavior for inf/NaN to integer conversions

Updates documentation in c10/util/TypeCast.h and torch/_tensor_docs.py
to specify that conversions from floating-point infinity and NaN to
integer types are now well-defined: +inf -> max, -inf -> min, NaN -> 0.
This behavior is consistent across CPU and CUDA devices.

Addresses documentation requirement for PR #162593.

vishalgoyal316 commented:
@pytorchbot rebase
